src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/TokenCertificatesApiController.java

```diff
@@ -0,0 +1,95 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.niis.xroad.restapi.openapi;
+
+import ee.ria.xroad.common.conf.serverconf.model.CertificateType;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.IOUtils;
+import org.niis.xroad.restapi.converter.CertificateDetailsConverter;
+import org.niis.xroad.restapi.exceptions.ErrorDeviation;
+import org.niis.xroad.restapi.openapi.model.CertificateDetails;
+import org.niis.xroad.restapi.service.CertificateAlreadyExistsException;
+import org.niis.xroad.restapi.service.ClientNotFoundException;
+import org.niis.xroad.restapi.service.GlobalConfService;
+import org.niis.xroad.restapi.service.KeyNotFoundException;
+import org.niis.xroad.restapi.service.TokenCertificateService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.Resource;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * certificates api
+ */
+@Controller
+@RequestMapping("/api")
+@Slf4j
+@PreAuthorize("denyAll")
+public class TokenCertificatesApiController implements TokenCertificatesApi {
+    public static final String ERROR_INVALID_CERT_UPLOAD = "invalid_cert_upload";
+    public static final String ERROR_INVALID_CERT = "invalid_cert";
+
+    private final TokenCertificateService tokenCertificateService;
+    private final CertificateDetailsConverter certificateDetailsConverter;
+
+    @Autowired
+    public TokenCertificatesApiController(TokenCertificateService tokenCertificateService,
+            CertificateDetailsConverter certificateDetailsConverter) {
+        this.tokenCertificateService = tokenCertificateService;
+        this.certificateDetailsConverter = certificateDetailsConverter;
+    }
+
+    @Override
+    @PreAuthorize("hasAnyAuthority('IMPORT_AUTH_CERT', 'IMPORT_SIGN_CERT')")
+    public ResponseEntity<CertificateDetails> addCertificate(Resource certificateResource) {
+        byte[] certificateBytes;
+        try (InputStream is = certificateResource.getInputStream()) {
+            certificateBytes = IOUtils.toByteArray(is);
+        } catch (IOException ex) {
+            throw new BadRequestException("cannot read certificate data", ex,
+                    new ErrorDeviation(ERROR_INVALID_CERT_UPLOAD));
+        }
+        CertificateType certificate = null;
+        try {
+            certificate = tokenCertificateService.addCertificate(certificateBytes);
+        } catch (GlobalConfService.GlobalConfOutdatedException | ClientNotFoundException | KeyNotFoundException
+                | TokenCertificateService.WrongCertificateUsageException
+                | TokenCertificateService.CsrNotFoundException
```

Comment:
Could CsrNotFoundException be more like ConflictException? That's what I thought initially, "csr does not exist" seems similar to "certificate already exists"? But I am not sure about this.

Reply:
Changed to `ConflictException`. Similar approach was used in [`TokensApiController#addKey`](https://github.com/nordic-institute/X-Road/blob/develop/src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/TokensApiController.java#L160)